### ISSUE TITLE
fitz.open blocks main thread even though I use it in a thread

### ISSUE THREAD

**Issue by yinkaisheng** [pymupdf/PyMuPDF#294]
Hi Jorj
I find that fitz.open costs a long time(more than 5 seconds) for some small epub files and the UI freezes for a few seconds.
So I use fitz.open in a thread and display the pages when it finishes opening, but the UI still freezes.

The following epub file and code can reproduce the issue.
[Python Crash Course(2nd).zip](https://github.com/pymupdf/PyMuPDF/files/3152179/Python.Crash.Course.2nd.zip)

```python
#!python3
# -*- coding: utf-8 -*-
import sys
import time
from threading import Thread
import fitz

def main(path):
    thread = Thread(target=thread_sleep)
    thread.start()
    while thread.is_alive():
        print('{}{}'.format('\b' * 20, time.perf_counter()), end='')
        sys.stdout.flush()
        time.sleep(0.01)
    print('\nrun open thread')
    thread = Thread(target=thread_open, args=(path, ))
    thread.start()
    while True:
        print('{}{}'.format('\b' * 20, time.perf_counter()), end='')
        sys.stdout.flush()
        time.sleep(0.01)

def thread_sleep():
    time.sleep(1)

def thread_open(path):
    start = time.perf_counter()
    doc = fitz.open(path)
    end = time.perf_counter()
    print('\nfitz.open {} costs {:.2f}s, it as {} pages'.format(path, end - start, doc.pageCount))
    doc.close()

if __name__ == '__main__':
    main(sys.argv[1])
```

![mupdfblocks](https://user-images.githubusercontent.com/5846696/57294259-a346bb00-70f9-11e9-8520-4c1e4ae5fdd2.gif)

the time doesn't update during fit.open

**Comment by JorjMcKie** [comment]
Does SumtraPDF perform ok?

Can you please try this document with the MuPDF's own native viewer as well? It is contained in the Windows download [here](https://mupdf.com/downloads/index.html).

**Comment by JorjMcKie** [comment]
Forget it. I tried both myself:
SumatraPDF performs ok and MuPDF v1.15 take a very long time and then doesn't show the first page.

**Comment by yinkaisheng** [comment]
Yes.
MuPDF also takes about 8s for the epub.
I don't care the open time now.
The key point is why it blocks main thread even fitz.open runs in a new thread.
It should not block main thread.

**Comment by JorjMcKie** [comment]
Hm, cannot answer this out of the hip. But note that (Py-) MuPDF is **not** threadable ...
I never have tried to use it under the `multiprocessing` umbrella instead, though.

**Comment by JorjMcKie** [comment]
It seems the main thread will not be blocked, when using the `multiprocessing` module - confirming my previous hypothesis:

```python
# -*- coding: utf-8 -*-
import sys
import time
import multiprocessing as mp
import fitz


def main(path):
    ctx = mp.get_context("spawn")
    q = ctx.Queue()
    p1 = ctx.Process(target=thread_sleep)
    p2 = ctx.Process(target=thread_open, args=(path,))
    p1.start()
    while p1.is_alive():
        print("{}{}".format("\b" * 20, time.perf_counter()), end="")
        sys.stdout.flush()
        time.sleep(0.01)
    print("\nrun open thread")
    p2.start()
    while True:
        print("{}{}".format("\b" * 20, time.perf_counter()), end="")
        sys.stdout.flush()
        time.sleep(0.01)


def thread_sleep():
    time.sleep(1)


def thread_open(path):
    start = time.perf_counter()
    doc = fitz.open(path)
    end = time.perf_counter()
    print(
        "\nfitz.open {} costs {:.2f}s, it has {} pages".format(
            path, end - start, doc.pageCount
        )
    )
    doc.close()
    print("after document close")


if __name__ == "__main__":
    main("Python-Crash-Course.epub")
```

**Comment by JorjMcKie** [comment]
btw, I opened that file in the cli mutool via ``mutool draw -o test.pdf python-crash-course.epub`` (to try convert it to a PDF). This exhibited a load of error messages like these:

```
D:\Jorj\Desktop>mutool draw -o test.pdf python-crash-course.epub 5
warning: table-cell not inside table-row element
error: css syntax error: expected number (OEBPS/Styles/Style00.css:7)
warning: ignoring stylesheet OEBPS/Styles/Style00.css
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: expected value (OEBPS/Styles/Style01.css:1)
error: css syntax error: expected value (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: expected value (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: expected number (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:1)
error: css syntax error: unexpected token (OEBPS/Styles/Style01.css:2)
error: css syntax error: expected value (OEBPS/Styles/Style01.css:3)
error: css syntax error: expected value (OEBPS/Styles/Style01.css:3)
... and 4000 errors more ...
```
Whether or not these messages are reporting true errors: they are the main reason for the extended open duration.

**Comment by JorjMcKie** [comment]
I also tried to convert this e-book to other formats and back to EPUB using `calibre`. After creating an AZW3 version and a new EPUB version from that, the file was quite quick to open with (Py-) MuPDF and "only" issued a few dozens of warnings and errors.
The epub file size also decreasing dramatically to about 1.5 MB (from 5.5 MB).
The issue with not showing the cover page remained however ...

**Comment by mozbugbox** [comment]
see swig doc:

http://www.swig.org/Doc4.0/Python.html#Python_multithreaded


    By default, SWIG does not enable support for
    multithreaded Python applications. More specifically,
    the Python wrappers generated by SWIG will not release
    the Python's interpreter's Global Interpreter Lock (GIL)
    when wrapped C/C++ code is entered. Hence, while any of
    the wrapped C/C++ code is executing, the Python
    interpreter will not be able to run any other threads,
    even if the wrapped C/C++ code is waiting in a blocking
    call for something like network or disk IO. 

You can try generate the swig wrapper with the `-threads` option

**Comment by JorjMcKie** [comment]
@mozbugbox, @yinkaisheng 
No, threads do simply not work with MuPDF like that.

The MuPDF internals use a "global context - 'gctx'". It serves as a central accelerator buffer for everything MuPDF does. This context area **_cannot be shared_** between threads. Or in other words: separate threads need separate context areas.

Everything happening in PyMuPDF is using **_one_** gctx.

We have been discussing this subject in detail quite some time ago ... you may want to dig out these earlier chats.

Bottom line: supporting threads would simply mean rewriting PyMuPDF. So we decided to not doing this.

As indicated in previous posts, multiprocessing remains as the only option.

**Comment by mozbugbox** [comment]
The thing with the current setup is that mupdf calls will block. Even if you dedicate a `mupdf manage thread` which handles all the mupdf (hence pymupdf) calls, that thread block the other none-mupdf threads because of the GIL.

We are not talking about multi-thread calls into mupdf randomly which will indeed mess up the none threadsafe global context.

**Comment by JorjMcKie** [comment]
@yinkaisheng - did this conversation answer your questions?

@yinkaisheng @mozbugbox - notwithstanding the foregoing:
It is **still possible** to create special-purpose functions that work with true threading support and MuPDF!
For example rendering all pages of a document (eg. as thumbnails) and storing ayway the results at some given location (folder).
You would have to provide the required parameters (folder, image resolution, desired page range, ...) and the function would create gctx copies and do the rendering in multi-threading mode. All this outside the Python level - obviously no Python calls occurring inside the function.

**Comment by yinkaisheng** [comment]
Yes.
I'll use multiprocessing.

**Comment by yinkaisheng** [comment]
Unfortunately, doc seems cannot be directly shared between two processes.
I put the doc in a multiprocessing.Queue, but the main process can't get the doc from the Queue.
Queue.qsize() returns 1 when I call it in the main process, but Queue.get() blocks forever.
Finally I made a workaround by using two Queues. One Queue for sending page numbers to the spawned process and one Queue for sending page pixmap bytes to the main process from the spawned process.
It worked.
![DocPlayer](https://user-images.githubusercontent.com/5846696/57427582-16b20f00-7257-11e9-8343-cd79aa18fa06.gif)

```Python
#!python3
# -*- coding: utf-8 -*-
# this script demostrates how to use PyMuPDF in multiprocessing to avoid unresponsive GUI when fitz.open costs a long time
import os
import sys
import time
import multiprocessing as mp
import queue
import fitz
from PyQt5 import QtCore, QtGui, QtWidgets

class DocForm(QtWidgets.QWidget):
    def __init__(self):
        super().__init__()
        self.process = None
        self.queNum = mp.Queue()
        self.queDoc = mp.Queue()
        self.pageCount = 0
        self.curPageNum = 0
        self.lastDir = ''
        self.timerSend = QtCore.QTimer(self)
        self.timerSend.timeout.connect(self.onTimerSendPageNum)
        self.timerGet = QtCore.QTimer(self)
        self.timerGet.timeout.connect(self.onTimerGetPage)
        self.timerWaiting = QtCore.QTimer(self)
        self.timerWaiting.timeout.connect(self.onTimerWaiting)
        self.initUI()

    def initUI(self):
        vbox = QtWidgets.QVBoxLayout()
        self.setLayout(vbox)

        hbox = QtWidgets.QHBoxLayout()
        self.btnOpen = QtWidgets.QPushButton('OpenDocument', self)
        self.btnOpen.clicked.connect(self.openDoc)
        hbox.addWidget(self.btnOpen)

        self.btnPlay = QtWidgets.QPushButton('PlayDocument', self)
        self.btnPlay.clicked.connect(self.playDoc)
        hbox.addWidget(self.btnPlay)

        self.btnStop = QtWidgets.QPushButton('Stop', self)
        self.btnStop.clicked.connect(self.stopPlay)
        hbox.addWidget(self.btnStop)

        self.label = QtWidgets.QLabel('0/0', self)
        hbox.addWidget(self.label)

        vbox.addLayout(hbox)

        self.labelImg = QtWidgets.QLabel('Document', self)
        self.label.setFont(QtGui.QFont('Verdana', 20))
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Preferred, QtWidgets.QSizePolicy.Expanding)
        self.labelImg.setSizePolicy(sizePolicy)
        vbox.addWidget(self.labelImg)

        self.setGeometry(100, 100, 400, 600)
        self.setWindowTitle('PyMuPDF Document Player')
        self.show()

    def openDoc(self):
        path, _ = QtWidgets.QFileDialog.getOpenFileName(self, "Open Document", self.lastDir,
                    "All Supported Files (*.pdf;*.epub;*.xps;*.oxps;*.cbz;*.fb2);;PDF Files (*.pdf);;EPUB Files (*.epub);;XPS Files (*.xps);;OpenXPS Files (*.oxps);;CBZ Files (*.cbz);;FB2 Files (*.fb2)", options=QtWidgets.QFileDialog.Options())
        if path:
            self.lastDir, self.file = os.path.split(path)
            if self.process:
                self.queNum.put(-1) # use -1 to notify the process to exit
            self.timerSend.stop()
            self.curPageNum = 0
            self.pageCount = 0
            self.process = mp.Process(target=openDocInProcess, args=(path, self.queNum, self.queDoc))
            self.process.start()
            self.timerGet.start(40)
            self.label.setText('0/0')
            self.queNum.put(0)
            self.startTime = time.perf_counter()
            self.timerWaiting.start(40)

    def playDoc(self):
        self.timerSend.start(500)

    def stopPlay(self):
        self.timerSend.stop()

    def onTimerSendPageNum(self):
        if self.curPageNum < self.pageCount - 1:
            self.queNum.put(self.curPageNum + 1)
        else:
            self.timerSend.stop()

    def onTimerGetPage(self):
        try:
            ret = self.queDoc.get(False)
            if isinstance(ret, int):
                self.timerWaiting.stop()
                self.pageCount = ret
                self.label.setText('{}/{}'.format(self.curPageNum + 1, self.pageCount))
            else:#tuple, pixmap info
                num, samples, width, height, stride, alpha = ret
                self.curPageNum = num
                self.label.setText('{}/{}'.format(self.curPageNum + 1, self.pageCount))
                fmt = QtGui.QImage.Format_RGBA8888 if alpha else QtGui.QImage.Format_RGB888
                qimg = QtGui.QImage(samples, width, height, stride, fmt)
                self.labelImg.setPixmap(QtGui.QPixmap.fromImage(qimg))
        except queue.Empty as ex:
            pass

    def onTimerWaiting(self):
        self.labelImg.setText('Loading "{}", {:.2f}s'.format(self.file, time.perf_counter() - self.startTime))

    def closeEvent(self, event):
        self.queNum.put(-1)
        event.accept()

def openDocInProcess(path, queNum, quePageInfo):
    start = time.perf_counter()
    doc = fitz.open(path)
    end = time.perf_counter()
    quePageInfo.put(doc.pageCount)
    while True:
        num = queNum.get()
        if num < 0:
            break
        page = doc.loadPage(num)
        pix = page.getPixmap()
        quePageInfo.put((num, pix.samples, pix.width, pix.height, pix.stride, pix.alpha))
    doc.close()
    print('process exit')

if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    form = DocForm()
    sys.exit(app.exec_())

```

**Comment by JorjMcKie** [comment]
This is awesome!
Would you mind writing a Wiki page about this in this repo?
Or would you agree if I include this script in the  package's documentation? Referencing you of course, including your copyright respectively?

**Comment by yinkaisheng** [comment]
Yes. Of course you can.

**Comment by JorjMcKie** [comment]
Thanks a lot.
Want to send me a copyright statement that I should include in the script source?

**Comment by yinkaisheng** [comment]
Just my email: yinkaisheng@live.com
